src/core/thread/mle.cpp

```diff
@@ -915,22 +915,23 @@ void Mle::SetRloc16(uint16_t aRloc16)
     if (aRloc16 != oldRloc16)
     {
         otLogNoteMle("RLOC16 %04x -> %04x", oldRloc16, aRloc16);
-
-        // Clear cached CoAP with old RLOC source
-        if (oldRloc16 != Mac::kShortAddrInvalid)
-        {
-            Get<Tmf::Agent>().ClearRequests(mMeshLocal16.GetAddress());
-        }
     }
 
-    Get<ThreadNetif>().RemoveUnicastAddress(mMeshLocal16);
+    if (Get<ThreadNetif>().HasUnicastAddress(mMeshLocal16) &&
+        (mMeshLocal16.GetAddress().GetIid().GetLocator() != aRloc16))
+    {
+        Get<ThreadNetif>().RemoveUnicastAddress(mMeshLocal16);
+        Get<Tmf::Agent>().ClearRequests(mMeshLocal16.GetAddress());
```

Comment:
Maybe beyond the scope of this PR, should we call `Get<Tmf::Agent>().ClearRequests` for other addresses (e.g. ALOC)? 
Maybe we should move `Get<Tmf::Agent>().ClearRequests` call to be inside Get<ThreadNetif>().RemoveUnicastAddress?

Reply:
@simonlingoogle it's a good idea, but I agree, better to do it in a separate PR.

Reply:
I agree on moving `ClearRequests()` and in a separate PR.